### PR TITLE
notification/feature/enhancments

### PR DESCRIPTION
**Pull request by devsadeq**

### CODE REVIEW

File: service_identity/local.properties [hunk 1]
```diff
@@ -4,5 +4,5 @@
 # Location of the SDK. This is only used by Gradle.
 # For customization when using a Version Control System, please read the
 # header note.
-#Tue Aug 01 11:49:29 AST 2023
-sdk.dir=/Users/ayamunadhil/Library/Android/sdk
+#Sun Aug 13 16:41:00 AST 2023
+sdk.dir=C\:\\Users\\devsa\\AppData\\Local\\Android\\Sdk
```
Comment (Mohammed-Alhams) [inline service_identity/local.properties:8]: Ignore this file or reomve it😬

**Review by HaidyAbuGom3a — APPROVED** [review]

File: service_notification/src/main/kotlin/org/thechance/service_notification/di/NotificationModule.kt [hunk 2]
```diff
@@ -17,11 +17,11 @@ val kmongoModule = module {
...
         val password = System.getenv("password")
-        KMongo.createClient("mongodb+srv://$username:$password@cluster0.0k23cqq.mongodb.net/")
+        KMongo.createClient("mongodb+srv://$username:$password@$cluster.mongodb.net/")
     }
 
     single {
-        get<MongoClient>().coroutine.getDatabase("TheChanceBeepBeep")
+        get<MongoClient>().coroutine.getDatabase("BeepBeepNotificationsDB")
     }
 }
```
Comment (Mohammed-Alhams) [inline service_notification/src/main/kotlin/org/thechance/service_notification/di/NotificationModule.kt:26]: I suggest you use the injection method used in the identity-service project

**Review by ay0o5h — APPROVED** [review]
LGTM

File: service_notification/src/main/kotlin/org/thechance/service_notification/domain/usecases/INotificationManagementUseCase.kt [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+                    body = body,
+                    date = System.currentTimeMillis(),
+                    userId = userId
+                )
+            )
+        }
+    }
+
+    override suspend fun sendNotificationToTopic(topic: String, title: String, body: String): Boolean {
+        if (!databaseGateway.isTopicAlreadyExists(topic)) throw InternalServerErrorException(TOPIC_NOT_EXISTS)
```
Comment (Mohammed-Alhams) [inline service_notification/src/main/kotlin/org/thechance/service_notification/domain/usecases/INotificationManagementUseCase.kt:39]: Throw Not found exception, not an internal server error if the resource doesn't exist

File: service_notification/src/main/kotlin/org/thechance/service_notification/domain/usecases/ITopicManagementUseCase.kt [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    suspend fun unsubscribeFromTopic(topic: String, token: String): Boolean
+}
+
+@Single
+class TopicManagementUseCase(
+    private val pushNotificationGateway: IPushNotificationGateway,
+    private val databaseGateway: IDatabaseGateway,
+) : ITopicManagementUseCase {
+    override suspend fun createTopic(topic: String): Boolean {
+        if (databaseGateway.isTopicAlreadyExists(topic)) throw InternalServerErrorException(ALREADY_EXISTS_TOPIC)
```
Comment (Mohammed-Alhams) [inline service_notification/src/main/kotlin/org/thechance/service_notification/domain/usecases/ITopicManagementUseCase.kt:26]: Here if the resource already exist, you have to throw ResourceAlreadyExistException, with status code 409(Conflict)

File: service_notification/src/main/kotlin/org/thechance/service_notification/domain/usecases/ITopicManagementUseCase.kt [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        return databaseGateway.createTopic(topic)
+    }
+
+    override suspend fun getTopics(): List<String> {
+        return databaseGateway.getTopics()
+    }
+
+    override suspend fun subscribeToTopic(topic: String, token: String): Boolean {
+        if (!databaseGateway.isTopicAlreadyExists(topic)) throw InternalServerErrorException(TOPIC_NOT_EXISTS)
```
Comment (Mohammed-Alhams) [inline service_notification/src/main/kotlin/org/thechance/service_notification/domain/usecases/ITopicManagementUseCase.kt:35]: NotFoundException, make one custom in the domain entites

File: service_notification/src/main/kotlin/org/thechance/service_notification/plugins/StatusPage.kt [hunk 6]
```diff
@@ -23,4 +24,8 @@ fun StatusPagesConfig.handleExceptions() {
         call.respond(HttpStatusCode.NotFound, listOf(cause.message?.toInt()))
     }
 
+    exception<InternalServerErrorException> { call, cause ->
+        call.respond(HttpStatusCode.InternalServerError, listOf(cause.message?.toInt()))
+    }
+
```
Comment (Mohammed-Alhams) [inline service_notification/src/main/kotlin/org/thechance/service_notification/plugins/StatusPage.kt:30]: To cut the duplication for exception handlers, you can check the identity-service

**Review by Mohammed-Alhams — APPROVED** [review]

**Review by Mohammed-Alhams — CHANGES_REQUESTED** [review]
Review the comments please

**Pull request closed without merging**